crates/polars-stream/src/nodes/shift.rs

```diff
@@ -0,0 +1,336 @@
+
+                        let tail = morsel.map(|df| df.tail(Some(len - shifted)));
+                        buffer.push_front(tail);
+
+                        morsel = shifted_morsel;
+                        shifted = 0;
+                    }
+                    // Return full morsel of nulls
+                    else {
+                        shifted -= len;
+                        let nulls = morsel
+                            .clone()
+                            .map(|df| DataFrame::full_null(&self.output_schema, df.height()));
+                        buffer.push_front(morsel);
+
+                        morsel = nulls;
+                    }
+
+                    if send_morsel(morsel, sender, seq).await.is_err() {
+                        break;
+                    }
+                }
+
+                let tail_shift = offset;
+                loop {
+                    if buffer.len_after_pop_front() >= tail_shift {
+                        let morsel = buffer.pop_front();
+
+                        if send_morsel(morsel, sender, seq).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        let Ok(morsel) = self.recv(receiver, state).await else {
+                            break;
+                        };
+                        buffer.push_back(morsel?);
+                    }
+                }
+
+                if buffer.len() != tail_shift {
+                    assert!(buffer.len_after_pop_front() < tail_shift);
+                    let last_morsel = buffer.pop_front();
+                    let len = tail_shift - buffer.len();
+                    let last_morsel = last_morsel.map(|df| df.slice(0, len));
+                    let _ = send_morsel(last_morsel, sender, seq).await;
+                }
+            },
+            State::Negative { skip, tail, seq } => {
+                while let Ok(morsel) = self.recv(receiver, state).await {
+                    let morsel = morsel?;
+
+                    if *skip == 0 {
+                        if send_morsel(morsel, sender, seq).await.is_err() {
+                            return Ok(());
+                        };
+                    } else {
+                        let len = morsel.df().len();
+
+                        if *skip >= len {
+                            *skip -= len
+                        } else {
+                            let morsel = morsel.map(|df| df.slice(*skip as _, usize::MAX));
+
+                            if send_morsel(morsel, sender, seq).await.is_err() {
+                                return Ok(());
+                            };
+                            *skip = 0;
+                        }
+                    }
+                }
+
+                let last_morsel = Morsel::new(
+                    DataFrame::full_null(&self.output_schema, *tail),
+                    Default::default(),
+                    SourceToken::new(),
```

Comment:
This is the last morsel that will send a morsel full of nulls. Am I allowed to make a `SourceToken` here?